# Avaliação Intermediária

Esta é a A.I. de Álgebra Linear e Teoria da Informação de 1s2023.



INSTRUÇÕES:

1. Faça todas as questões
1. Salve o notebook
1. Envie o notebook salvo no Blackboard

IMPORTANTE:

* Você pode consultar qualquer material ou IA, mas não qualquer colega ou pessoa humana, presente ou não na prova.
* O envio da prova será fechado às 17h45. Aqueles que não enviarem o material até este horário ficarão com zero.
* A prova vale um total de 10 pontos. As questões 1, 2, 3 e 4 têm, respectivamente, pesos 1, 2, 3 e 4.
* Os conceitos F, E, D, C, B, A serão mapeados para 0, 2, 4, 6, 8 e 10.
* Para todas as questões, usaremos a rubrica que segue anexa a esta prova. Há uma rubrica para a justificativa e uma rubrica para o código. A nota da questão é a menor entre as notas da justificativa e do código.


In [9]:
import numpy as np

## Exercício 1

* Peso: 1 ponto

Dadas as matrizes:

$$
A = \begin{bmatrix} 1 & 3 &\\ 2 & 1 \end{bmatrix} \\

B = \begin{bmatrix} 2 & 3\\ 5 & 8 \end{bmatrix} \\

C = AB \\

D = BA 
$$

calcule:

$X = (ABC^{-1}BAD^{-1})^T \odot A$



In [10]:
A = np.array([[1, 3], [2, 1]])
B = np.array([[2, 3], [5, 8]])

C = A @ B
D = B @ A

X = (A @ B @ np.linalg.inv(C) @ B @ A @ np.linalg.inv(D)).T * A

X

array([[1.00000000e+00, 1.33226763e-14],
       [1.19015908e-14, 1.00000000e+00]])

# Exercício 2

* Peso: 2 pontos

Tenho uma matriz $X$ que contém todas as posições $(i,j)$ de pixels de uma figura de 600x600 pixels, sendo cada pixel armazenado em uma coluna e cada dimensão armazenada em uma linha. Qual é a matriz que permite encolher essa figura de forma que cada uma de suas dimensões seja reduzida pela metade, mas seu centro seja preservado? Se precisar, use coordenadas homogêneas!

________________________________________________________________________________________________________

$$
translacao para ponto (0,0) = 
\begin{bmatrix}
    1 & 0 & -300 \\
    0 & 1 & -300 \\
    0 & 0 & 1
\end{bmatrix}
$$


$$
compressaometade = 
\begin{bmatrix}
    0.5 & 0 & 0 \\
    0 & 0.5 & 0 \\
    0 & 0 & 1
\end{bmatrix}
$$

$$
translacao para ponto original = 
\begin{bmatrix}
    1 & 0 & 300 \\
    0 & 1 & 300 \\
    0 & 0 & 1
\end{bmatrix}
$$

In [11]:
translacao_1 = np.array([[1, 0, -300], [0, 1, -300], [0, 0, 1]])

compressao = np.array([[0.5, 0, 0], [0, 0.5, 0], [0, 0, 1]])

translacao_2 = np.array([[1, 0, 300], [0, 1, 300], [0, 0, 1]])

M = translacao_2 @ compressao @ translacao_1

M

array([[  0.5,   0. , 150. ],
       [  0. ,   0.5, 150. ],
       [  0. ,   0. ,   1. ]])

# Exercício 3

* Peso: 3 pontos

Algumas carteiras de investimento deram os seguintes retornos:

* 100 ações da empresa E1 e 200 ações da E2 deram retorno de R$100,00,
* 1000 ações da empresa E1 e 2000 ações da E2 deram retorno de R$1000,00,
* 200 ações da empresa E2 e 300 da E3 deram retorno de R$500,00 
* 400 ações da empresa E2 e 600 da E3 deram retorno de R$1000,00 
* 1000 ações da empresa E1 e 100 da E3 deram retorno de R$1500,00
* 2000 ações da empresa E1 e 200 da E3 deram retorno de R$3000,00 

*Usando estritamente formulações matriciais*, responda:
1. Qual é o retorno por ação de cada empresa?
1. Qual é o retorno de uma carteira com 100 ações da empresa E1, 200 da empresa E2 e 300 da E3?

________________________________________________________________________________________________________

1 e 2, 3 e 4, 5 e 6, essas duplas sao carteiras que possuem mesmas acoes apenas em quantidades diferentes, ou seja, possuem a mesma informacao.
Por isso considerarei apenas as carteiras 1, 3 e 5

In [15]:
c1 = np.array([100, 200, 0])
c2 = np.array([0, 200, 300])
c3 = np.array([1000, 0, 100])

c = np.vstack((c1, c2, c3))

r = np.array([[100], [500], [1500]])

c: carteiras
x: matriz_retornos_por_acao
r: retornos_totais

$$cx = r$$
$$c^{-1}cx = c^{-1}r$$
$$x = c^{-1}r$$


In [16]:
X = np.linalg.inv(c) @ r
X

array([[ 1.32258065],
       [-0.16129032],
       [ 1.77419355]])

1. O retorno da acao E1 é de 1.32258065 vezes o seu valor, da E2 é de -0.16129032 e da E3 de 1.77419355.

In [17]:
c_nova = np.array([100, 200, 300])
retorno_total = c_nova @ X
retorno_total

array([632.25806452])

632.25806452 reais

# Exercício 4

* Peso: 4 pontos

Uma maneira de fazer transformações em imagens é usar pontos de referência. Por exemplo, podemos definir pontos de referência em um rosto como o centro de cada um dos olhos, a ponta do nariz e o ponto inferior do queixo. Daí então, encontramos onde esses pontos estão na imagem de origem, e escolhemos (como quisermos) onde os pontos estão na imagem de destino. Encontramos a transformação linear que realiza esse mapeamento, e então aplicamos essa mesma transformação para todos os pontos da imagem, isto é:

$$
Xd = T X,
$$

onde $Xd$ é uma matriz com pontos na imagem de destino, $X$ é uma matriz com pontos na imagem de origem, e $T$ é uma transformação linear.

Num rosto, em uma imagem que está mapeada no plano cartesiano (isto é, no plano usual $x,y$ e não numa matriz de pixels), encontramos:

* O olho esquerdo em $(-2, 3)$,
* o olho direito em $(2, 3)$
* A ponta do nariz em $(0,0)$
* O ponto inferior do queixo em $(0, -3)$

O rosto foi mapeado usando uma transformação linear de forma que os pontos de referência foram deslocados para:

* O olho esquerdo em $(-6, 2)$,
* o olho direito em $(6, 1)$
* A ponta do nariz em $(3, 0)$

Usando formulações matriciais, responda: qual é a distância entre o ponto inferior do queixo e a ponta do nariz no novo mapeamento?


________________________________________________________________________________________________________

Primeiro precisamos encontrar a matriz que faz as seguintes transformacoes:

* $[-2, 3] \rightarrow [-6, 2]$
* $[2, 3] \rightarrow [6, 1]$
* $[0, 0] \rightarrow [3, 0]$

Para levar o ponto (0, -3) para seu destino e em seguida calcular a distancia entre esse novo ponto e (3, 0)



In [19]:
v1 = np.array([-2, 3])
v2 = np.array([2, 3])
v3 = np.array([0, 0])

v = np.vstack([v1, v2, v3]).T
v = np.vstack((v, np.ones(v.shape[1])))

w1 = np.array([-6, 2])
w2 = np.array([6, 1])
w3 = np.array([3, 0])

w = np.vstack([w1, w2, w3]).T
w = np.vstack((w, np.ones(w.shape[1])))

B = v @ np.linalg.inv(w)

B

array([[ 4.00000000e-01,  8.00000000e-01, -1.20000000e+00],
       [ 2.00000000e-01,  2.40000000e+00, -6.00000000e-01],
       [-1.38777878e-17, -2.22044605e-16,  1.00000000e+00]])

In [21]:
ponto_queixo = np.array([[0], [-3], [1]])
ponto_novo_queixo = np.linalg.inv(B) @ ponto_queixo

ponto_novo_queixo

array([[ 6. ],
       [-1.5],
       [ 1. ]])

O ponto destino do queixo sera (6, -1.5), agora calculamos a distancia entre ele e (3, 0)

In [24]:
distancia_2 = ((6 - 3) ** 2) + ((-1.5 - 0) ** 2)
distancia = distancia_2 ** 0.5
distancia

3.3541019662496847